src/realm/sync/noinst/client_history_impl.cpp

```diff
@@ -454,12 +454,15 @@ void ClientHistory::integrate_server_changesets(
             update_sync_progress(progress, downloadable_bytes, transact); // Throws
         }
         // Always update progress for download messages from steady state.
-        else if (batch_state == DownloadBatchState::SteadyState) {
+        else if (batch_state == DownloadBatchState::SteadyState && !changesets_to_integrate.empty()) {
             auto partial_progress = progress;
             partial_progress.download.server_version = last_changeset.remote_version;
```

Comment:
You could maybe simplify this a bit by moving the `changesets_to_integrate.empty()` check to here? So

```
else if (batch_state == DownloadBatchState::SteadyState) {
    auto partial_progress = progress;
    // Comment about why we have to do this futzing with cursors here.
    if (!changesets_to_integrate.empty()) {
        partial_progress.download.server_version = last_changeset.remote_version;
        ...
    }

    update_sync_progress(partial_progress, ...);
}
```

Reply:
My first implementation was like that but I didn't like it :) The idea being that it's not really a partial progress anymore (that happens both when all changesets are integrated in one go, or when the last chunk is integrated if a split was needed), so I wanted it's own branch. No strong opinion though.